core/modules/utils/linked-list.js

```diff
@@ -14,105 +14,181 @@ function LinkedList() {
-	for(var ptr = this.next; ptr !== this; ptr = ptr.next) {
-		callback(ptr.value);
+	var visits = Object.create(null),
+		value = this.first;
+	while(value !== undefined) {
+		callback(value);
+		var next = this.next[value];
+		if(typeof next === "object") {
+			var i = visits[value] || 0;
+			visits[value] = i+1;
+			value = next[i];
+		} else {
+			value = next;
+		}
 	}
 };
 
 LinkedList.prototype.toArray = function() {
 	var output = [];
-	for(var ptr = this.next; ptr !== this; ptr = ptr.next) {
-		output.push(ptr.value);
-	}
+	this.each(function(value) { output.push(value); });
 	return output;
 };
 
 function _removeOne(list,value) {
-	var node = list.index[value];
-	if(node) {
-		node.prev.next = node.next;
-		node.next.prev = node.prev;
-		list.length -= 1;
-		// Point index to the next instance of the same value, maybe nothing.
-		list.index[value] = node.copy;
+	var prevEntry = list.prev[value],
+		nextEntry = list.next[value],
+		prev = prevEntry,
+		next = nextEntry;
+	if(typeof nextEntry === 'object') {
+		next = nextEntry[0];
+		prev = prevEntry[0];
 	}
-	return node;
+	// Relink preceding element.
+	if(list.first === value) {
+		list.first = next
+	} else if(prev !== undefined) {
+		if(typeof list.next[prev] === 'object') {
+			if(next === undefined) {
+				// Must have been last, and 'i' would be last element.
+				list.next[prev].pop();
+			} else {
+				var i = list.next[prev].indexOf(value);
+				list.next[prev][i] = next;
+			}
+		} else {
+			list.next[prev] = next;
+		}
+	} else {
+		return;
+	}
+	// Now relink following element
+	// Check "next !== undefined" rather than "list.last === value" because
+	// we need to know if the FIRST value is the last in the list, not the last.
+	if(next !== undefined) {
+		if(typeof list.prev[next] === 'object') {
+			if(prev === undefined) {
+				// Must have been first, and 'i' would be 0.
+				list.prev[next].shift();
+			} else {
+				var i = list.prev[next].indexOf(value);
+				list.prev[next][i] = prev;
+			}
+		} else {
+			list.prev[next] = prev;
+		}
+	} else {
+		list.last = prev;
+	}
+	// Delink actual value. If it uses arrays, just remove first entries.
+	if(typeof nextEntry === 'object') {
+		nextEntry.shift();
+		prevEntry.shift();
+	} else {
+		list.next[value] = undefined;
+		list.prev[value] = undefined;
+	}
+	list.length -= 1;
 };
 
-function _linkToEnd(list,node) {
-	// Sticks the given node onto the end of the list.
-	list.prev.next = node;
-	node.prev = list.prev;
-	list.prev = node;
-	node.next = list;
+// Sticks the given node onto the end of the list.
+function _linkToEnd(list,value) {
+	if(list.first === undefined) {
+		list.first = value;
+	} else {
+		// Does it already exists?
+		if(list.first === value || list.prev[value] !== undefined) {
+			if(typeof list.next[value] === 'string') {
+				list.next[value] = [list.next[value]];
+				list.prev[value] = [list.prev[value]];
+			} else if(typeof list.next[value] === 'undefined') {
```

Comment:
Apologies, I may have missed this in earlier commits, but our coding style is to always use double quotes for string literals (there are some instances where we use single quotes for strings that themselves contain double quotes).